pennylane/dla/lie_closure.py

```diff
@@ -24,6 +26,136 @@
+
+    Let us walk through a simple example of computing the Lie closure of the generators of the transverse field Ising model on two qubits.
+
+    ```python
+    ops = [X(0) @ X(1), Z(0), Z(1)]
+    ```
+
+    A first round of commutators between all elements yields the new operators `Y(0) @ X(1)` and `X(0) @ Y(1)`.
+
+    ```python
+    >>> qml.commutator(X(0) @ X(1), Z(0))
+    2j * (Y(0) @ X(1))
+    >>> qml.commutator(X(0) @ X(1), Z(0))
+    2j * (X(0) @ Y(1))
+    ```
+
+    A next round of commutators between all elements further yields the new operator `Y(0) @ Y(1)`.
+
+    ```python
+    >>> qml.commutator(X(0) @ Y(1), Z(0))
+    -2j * (Y(0) @ Y(1))
+    ```
+
+    After that, no new operators emerge from taking nested commutators and we have the resulting DLA.
+    This can now be done in short via `qml.dla.lie_closure` as follows.
+
+    ```python
+    >>> ops = [X(0) @ X(1), Z(0), Z(1)]
+    >>> dla = qml.dla.lie_closure(ops)
+    >>> print(dla)
+    [X(1) @ X(0),
+     Z(0),
+     Z(1),
+     -1.0 * (X(1) @ Y(0)),
+     -1.0 * (Y(1) @ X(0)),
+     -1.0 * (Y(1) @ Y(0))]
+
+    .. details::
+        :title: Usage Details
+
+        Note that by default, ``lie_closure`` returns PennyLane operators. Internally we use the more
+        efficient representation in terms of :class:`~PauliSentence` by making use of the ``op.pauli_rep``
+        attribute of operators composed of Pauli operators. If desired, this format can be returned by using
+        the keyword ``pauli=True``. In that case, the input is also assumed to be a :class:`~PauliSentence` instance.
+
+        >>> ops = [
+        ...     PauliSentence({PauliWord({0: "X", 1: "X"}): 1.}),
+        ...     PauliSentence({PauliWord({0: "Z"}): 1.}),
+        ...     PauliSentence({PauliWord({1: "Z"}): 1.}),
+        ... ]
+        >>> dla = qml.dla.lie_closure(ops, pauli=True)
+        >>> print(dla)
+        [1.0 * X(0) @ X(1),
+         1.0 * Z(0),
+         1.0 * Z(1),
+         -1.0 * Y(0) @ X(1),
+         -1.0 * X(0) @ Y(1),
+         -1.0 * Y(0) @ Y(1)]
+        >>> type(dla[0])
+        pennylane.pauli.pauli_arithmetic.PauliSentence
+
+    """
+    # This check and conversion is ignored in pauli=True mode
+    if not pauli and not all(isinstance(op, PauliSentence) for op in generators):
+        generators = [
+            op.pauli_rep if op.pauli_rep else qml.pauli.pauli_sentence(op) for op in generators
+        ]
+
+    vspace = PauliVSpace(generators)
+
+    epoch = 0
+    old_length = 0  # dummy value
+    new_length = len(vspace)
+
+    while new_length > old_length or epoch > max_iterations:
+        if verbose > 0:
+            print(f"epoch {epoch+1} of lie_closure")
+        for ps1, ps2 in itertools.combinations(vspace.basis, 2):
+            com = ps1.commutator(ps2)
+            if len(com) == 0:  # skip because operators commute
+                continue
+
+            # result is always purely imaginary
+            # TODO potentially renormalize?
```

Comment:
Is this still a to do?

Reply:
Jein, I remove the factor 2 already, which gives neat results for all examples with pure pauli words.
**Could** still do:
 * Remove minus factors
 * normalize also for pauli sentences, but this is significantly more expensive

Reply:
Actually doesnt look too bad for pauli sentenes either:
```python
n = 4
generators = [
    PauliSentence({PauliWord({i: "X", (i + 1) % n: "X"}): 1.0, PauliWord({i: "Y", (i + 1) % n: "Y"}): 1.0}) for i in range(n - 1)
]
generators += [
    PauliSentence({PauliWord({i: "Z"}): 1.0}) for i in range(n)
]

res = qml.dla.lie_closure(generators)
res
```

```
[X(0) @ X(1) + Y(0) @ Y(1),
 X(1) @ X(2) + Y(1) @ Y(2),
 X(2) @ X(3) + Y(2) @ Y(3),
 Z(0),
 Z(1),
 Z(2),
 Z(3),
 Z(1) @ Y(2) @ X(0) + -1.0 * (Z(1) @ X(2) @ Y(0)),
 -1.0 * (Y(0) @ X(1)) + X(0) @ Y(1),
 Z(2) @ Y(3) @ X(1) + -1.0 * (Z(2) @ X(3) @ Y(1)),
 -1.0 * (Y(1) @ X(2)) + X(1) @ Y(2),
 -1.0 * (Y(2) @ X(3)) + X(2) @ Y(3),
 (
     -1.0 * (Z(2) @ X(3) @ Z(1) @ X(0))
   + -1.0 * (Z(2) @ Y(3) @ Z(1) @ Y(0))
 ),
 (
     -1.0 * (Z(1) @ X(2) @ X(0))
   + -1.0 * (Z(1) @ Y(2) @ Y(0))
 ),
 (
     -1.0 * (Z(2) @ X(3) @ X(1))
   + -1.0 * (Z(2) @ Y(3) @ Y(1))
 ),
 (
     -1.0 * (Z(1) @ Z(2) @ X(0) @ Y(3))
   + Z(1) @ Z(2) @ Y(0) @ X(3)
 )]
```

Reply:
Yup, I agree that this is as neat as it gets without doing manual work or adding unnecessary conventions :)

Reply:
Would suggest to remove the code comment, then :)